### PR TITLE
Update docs to new frontend/UI

### PR DESCRIPTION
**Pull request by mishig25**
### TLDR: Update `datasets` `docs` to the new syntax (markdown and mdx files) & frontend (as how it looks on [hf.co/transformers](https://huggingface.co/docs/transformers/index))

| Light mode                                                                                 | Dark mode                                                             |
|-----------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------|
| Screenshot 2022-02-17 at 14 15 34    | Screenshot 2022-02-17 at 14 16 27                                     |

## Checklist

- [x] update datasets docs to new syntax (should call `doc-builder convert`) (this PR)
- [x] discuss `@property` methods frontend https://github.com/huggingface/doc-builder/pull/87
- [x] discuss `inject_arrow_table_documentation` (this PR) https://github.com/huggingface/datasets/pull/3690#discussion_r801847860
- [x] update datasets docs path on moon-landing https://github.com/huggingface/moon-landing/pull/2089
- [x] convert pyarrow docstring from Numpydoc style to groups style  https://github.com/huggingface/doc-builder/pull/89(https://stackoverflow.com/a/24385103/6558628)
- [x] handle `Raises` section on frontend and doc-builder https://github.com/huggingface/doc-builder/pull/86
- [x] check imgs path (this PR) (nothing to update here)
- [x] doc exaples block has to follow format `Examples::` https://github.com/huggingface/datasets/pull/3693
- [x] fix [this docstring](https://github.com/huggingface/datasets/blob/6ed6ac9448311930557810383d2cfd4fe6aae269/src/datasets/arrow_dataset.py#L3339) (causing svelte compilation error)
- [x] Delete sphinx related files 
- [x] Delete sphinx CI
- [x]  Update docs config in setup.py
- [x] add `versions.yml` in doc-build  https://github.com/huggingface/doc-build/pull/1
- [x] add `versions.yml` in doc-build-dev https://github.com/huggingface/doc-build-dev/pull/1
- [x] https://github.com/huggingface/moon-landing/pull/2089
- [x] format docstrings for example `datasets.DatasetBuilder.download_and_prepare` args format look wrong
- [x] create new github actions. (can probably be in a separate PR) (see the transformers equivalents below)
1. [build_dev_documentation.yml](https://github.com/huggingface/transformers/blob/master/.github/workflows/build_dev_documentation.yml)
2. [build_documentation.yml](https://github.com/huggingface/transformers/blob/master/.github/workflows/build_documentation.yml)
3. [delete_dev_documentation.yml](https://github.com/huggingface/transformers/blob/master/.github/workflows/delete_dev_documentation.yml)

## Note to reviewers

The number of changed files is a lot (100+) because I've converted all `.rst` files to `.mdx` files & they are compiling fine on the svelte side (also, moved all the imgs to to [doc-imgs repo](https://huggingface.co/datasets/huggingface/documentation-images/tree/main/datasets)). Moreover, you should just review them on preprod and see if the rendering look fine.
_Therefore, I'd suggest to focus on the changed_ **`.py`** and **CI files** (github workflows, etc. you can use [this filter here](https://github.com/huggingface/datasets/pull/3690/files?file-filters%5B%5D=.py&file-filters%5B%5D=.yml&show-deleted-files=true&show-viewed-files=true)) during the review & ignore `.mdx` files. (if there's a bug in `.mdx` files, we can always handle it in a separate PR afterwards).

### CODE REVIEW

File: src/datasets/table.py [hunk 1]
```diff
@@ -24,10 +24,12 @@
 
 
 def inject_arrow_table_documentation(arrow_table_method):
-    def wrapper(method):
-        out = wraps(arrow_table_method)(method)
```
Comment (mishig25) [inline src/datasets/table.py:28]: Is it safe to replace `wraps` with manually setting `out.__doc__ = arrow_table_method.__doc__` ?
Otherwise, It was breaking the doc-builder's `inspect` usage.

Comment (mishig25) [reply src/datasets/table.py:28]: basically, are we using other properties set by `wraps` such as `__dict__`?

Comment (mishig25) [reply src/datasets/table.py:28]: @lhoestq @albertvillanova @mariosasko

Comment (lhoestq) [reply src/datasets/table.py:28]: It's useful to have the same typing as well via `__annotations__ `, but `__dict__` is probably not important.

Feel free to also set the `__annotations__` as well here !

Comment (mishig25) [reply src/datasets/table.py:28]: handled in https://github.com/huggingface/datasets/pull/3690/commits/865261eaecbfbc18e265e16e9f62ce3f26713dff

**Comment by lhoestq** [comment]
We can have the docstrings of the properties that are missing docstrings (from discussion [here](https://github.com/huggingface/doc-builder/pull/96)) here by using your new `inject_arrow_table_documentation` onthem as well ?

File: src/datasets/table.py [hunk 2]
```diff
@@ -384,7 +388,7 @@ def from_arrays(cls, *args, **kwargs):
         return cls(pa.Table.from_arrays(*args, **kwargs))
 
     @classmethod
-    @inject_arrow_table_documentation(pa.Table.from_pydict)
+    # @inject_arrow_table_documentation(pa.Table.from_pydict)
```
Comment (lhoestq) [inline src/datasets/table.py:391]: I think you can also uncomment this ?

**Comment by mishig25** [comment]
@sgugger & @lhoestq could you help me with what should the `docs` section in setup.py be changed to [here](https://github.com/huggingface/datasets/blob/master/setup.py#L212-L227) ?

as a reference, here is a transformers setup.py docs [section](https://github.com/huggingface/transformers/blob/master/setup.py#L304-L308)

**Comment by sgugger** [comment]
For now, you can put an empty list. Once the `doc-builder` is in a PyPi package (with the bug we fixed on Datasets but still waiting on the standing PR with the code switch) we can put it there.

**Comment by mishig25** [comment]
None of those dependencies are needed from this list?

```py
    "docs": [
        "docutils==0.16.0",
        "recommonmark",
        "sphinx==3.1.2",
        "sphinx-markdown-tables",
        "sphinx-rtd-theme==0.4.3",
        "sphinxext-opengraph==0.4.1",
        "sphinx-copybutton",
        "fsspec<2021.9.0",
        "s3fs",
        "sphinx-panels",
        "sphinx-inline-tabs",
        "myst-parser",
        "Markdown!=3.3.5",
    ],
```

**Comment by mishig25** [comment]
@lhoestq feel free to request reviews from other maintainers 😊

**Comment by lhoestq** [comment]
Thanks ! @mariosasko and @albertvillanova feel free to take a look :)
I can do a thorough review this afternoon

File: docs/source/process.mdx [hunk 3]
```diff
@@ -0,0 +1,627 @@
...
+>>> dataset = load_dataset('glue', 'mrpc', split='train')
+```
+
+<Tip warning={true}>
+
+All the processing methods in this guide return a new [`datasets.Dataset`]. Modification is not done in-place. Be careful about overriding your previous dataset!
+
+</Tip>
+
+## Sort, shuffle, select, split, and shard
```
Comment (lhoestq) [inline docs/source/process.mdx:27]: For some reason these section titles are not rendered correctly in the right bar:
image

**Review by stevhliu — COMMENTED** [review]
This looks so good, great job on the conversion! 🔥 

Left a few comments about the syntax, and text in the right navbar isn't being separated (ie Load metric looks like Loadmetric).

File: docs/source/package_reference/main_classes.mdx [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+## `Metric`
+
+The base class `Metric` implements a Metric backed by one or several [`datasets.Dataset`].
+
+[[autodoc]] datasets.Metric
+
+## `Filesystems`
+
+
+[[autodoc]] datasets.filesystems.S3FileSystem(anon=False, key=None, secret=None, token=None, use_ssl=True, client_kwargs=None, requester_pays=False, default_block_size=None, default_fill_cache=True, default_cache_type='bytes', version_aware=False, config_kwargs=None, s3_additional_kwargs=None, session=None, username=None, password=None, asynchronous=False, loop=None, **kwargs)
```
[the remark this answers is not in the document]
Comment (lhoestq) [reply docs/source/package_reference/main_classes.mdx:228]: Maybe the `s3fs` dependency is missing when building the documentation ?

This is an optional dependency and `S3FileSysytem` only exists if `s3fs` is installed

File: docs/source/access.mdx [hunk 5]
```diff
@@ -0,0 +1,126 @@
+# The `Dataset` object
+
+In the previous tutorial, you learned how to successfully load a dataset. This section will familiarize you with the [`datasets.Dataset`] object. You will learn about the metadata stored inside a Dataset object, and the basics of querying a Dataset object to return rows and columns.
```
[the remark this answers is not in the document]
Comment (mishig25) [reply docs/source/access.mdx]: also removed other instances https://github.com/huggingface/datasets/pull/3690/commits/239c8c166c64ef3d7ca07263bda2ab549053062b

File: docs/source/package_reference/main_classes.mdx [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+## `DatasetInfo`
+
+[[autodoc]] datasets.DatasetInfo
+
+## `Dataset`
+
+The base class [`datasets.Dataset`] implements a Dataset backed by an Apache Arrow table.
+
+[[autodoc]] datasets.Dataset
+    -  add_column
```
[the remark this answers is not in the document]
Comment (mishig25) [reply docs/source/package_reference/main_classes.mdx:13]: I think we can just use italics to write this info https://github.com/huggingface/datasets/pull/3690/commits/fae3af3e2935e1deb9058086d40c4098b8ac432a
Screenshot 2022-02-19 at 23 25 33

File: docs/source/package_reference/builder_classes.mdx [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+[[autodoc]] datasets.SplitGenerator
+
+[[autodoc]] datasets.Split
+
+[[autodoc]] datasets.NamedSplit
+
+[[autodoc]] datasets.NamedSplitAll
+
+[[autodoc]] datasets.ReadInstruction
```
[the remark this answers is not in the document]
Comment (mishig25) [reply docs/source/package_reference/builder_classes.mdx:27]: Thanks for the catch! https://github.com/huggingface/datasets/pull/3690/commits/ec7ac80b3e8f2ed806ba76d40cf6e975acbd0667
Screenshot 2022-02-19 at 23 30 17

File: docs/source/package_reference/main_classes.mdx [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+    - rename_column
+    - class_encode_column
+    - push_to_hub
+    - save_to_disk
+    - load_from_disk
+    - from_csv
+    - from_json
+    - from_parquet
+    - from_text
+    - prepare_for_task
```
[the remark this answers is not in the document]
Comment (mishig25) [reply docs/source/package_reference/main_classes.mdx:151]: taken care in https://github.com/huggingface/datasets/pull/3690/commits/0745014a3068ab72e9776cb916534afd22833411

File: docs/source/package_reference/main_classes.mdx (hunk 4 above)
[the remark this answers is not in the document]
Comment (mishig25) [reply docs/source/package_reference/main_classes.mdx]: thanks for the catch! handled in https://github.com/huggingface/datasets/pull/3690/commits/99dcf1695fe8d03614fec963f886aa88960b95c2 & https://github.com/huggingface/datasets/pull/3690/commits/88eec86c6667acca8e678963c282a9bb62ee5a87

File: docs/source/process.mdx (hunk 3 above)
Comment (mishig25) [reply docs/source/process.mdx:27]: https://github.com/huggingface/datasets/pull/3690/commits/eb7fecb0cf9ba61356758bc7316ff1bb150d8f59 Following the design choice we took on the transformers doc where we remove `<code>` format from the headings.
cc: @gary149 
Screenshot 2022-02-21 at 09 15 55

File: docs/source/about_dataset_load.mdx [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+## Building a dataset
+
+When you load a dataset for the first time, 🤗 Datasets takes the raw data file and builds it into a table of rows and typed columns. There are two main classes responsible for building a dataset: [`datasets.BuilderConfig`] and [`datasets.DatasetBuilder`].
+
+
+<div class="flex justify-center">
+   <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/datasets/builderconfig.png"/>
+</div>
+
+### [`datasets.BuilderConfig`]
```
[the remark this answers is not in the document]
Comment (mishig25) [reply docs/source/about_dataset_load.mdx]: https://github.com/huggingface/datasets/pull/3690/commits/c1bf3efec107cea5465ead219c0f49bedfb7c69a  Since `datasets.BuilderConfig` link is already the first word of the paragraph, I've removed the link from the heading (as related to https://github.com/huggingface/datasets/pull/3690#discussion_r811150659)
Screenshot 2022-02-21 at 14 49 30

**Comment by lhoestq** [comment]
Cool thanks ! Feel free to merge master into this branch and run `make style` to fix the python code formatting

File: docs/source/package_reference/builder_classes.mdx [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+
+[[autodoc]] datasets.BeamBasedBuilder
+
+[[autodoc]] datasets.ArrowBasedBuilder
+
+[[autodoc]] datasets.BuilderConfig
+
+[[autodoc]] datasets.DownloadManager
+
+[[autodoc]] datasets.GenerateMode
```
Comment (stevhliu) [inline docs/source/package_reference/builder_classes.mdx:17]: Display the table in Markdown in the docstring for `GenerateMode`:

|                                     | Downloads | Dataset |
|-------------------------------------|-----------|---------|
| `REUSE_DATASET_IF_EXISTS` (default) | Reuse     | Reuse   |
| `REUSE_CACHE_IF_EXISTS`             | Reuse     | Fresh   |
| `FORCE_REDOWNLOAD`                  | Fresh     | Fresh   |

**Review by stevhliu — COMMENTED** [review]
Just one last thing I noticed with a table not being properly displayed in the docstring!

**Comment by stevhliu** [comment]
Love the colorful vibes here!
Screen Shot 2022-02-22 at 9 54 17 AM

File: docs/source/package_reference/builder_classes.mdx (hunk 10 above)
Comment (mishig25) [reply docs/source/package_reference/builder_classes.mdx]: handled in https://github.com/huggingface/datasets/pull/3690/commits/1ff531cda64d1aa44230c277b5322051312f1816

**Comment by mishig25** [comment]
> Could you update preprod please ? Or is there a preview somewhere I can check to make sure everything is ok ?

I'll let you know once preprod gets updated

**Review by lhoestq — APPROVED** [review]
It looks all good ! Thanks a lot !

File: docs/source/audio_process.mdx [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+The [`datasets.Audio`] feature is an experimental feature and should be installed as an extra dependency in 🤗 Datasets. Install the [`datasets.Audio`] feature with pip:
+
+```
+>>> pip install datasets[audio]
+```
+
+Users should also install [torchaudio](https://pytorch.org/audio/stable/index.html) and [librosa](https://librosa.org/doc/latest/index.html), two common libraries used by 🤗 Datasets for handling audio data.
+
+```
+>>> pip install librosa
```
[the remark this answers is not in the document]
Comment (mishig25) [reply docs/source/audio_process.mdx:20]: handled in https://github.com/huggingface/datasets/pull/3690/commits/e60d51af0f9087484c83d9e41cb59f0ce0214f76

thanks a lot for catching it!

File: .github/workflows/build_documentation.yml [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+          cd ..
+
+          cd notebooks
+          git pull origin master
+          cd ..
+
+      - name: Make documentation
+        run: |
+          cd doc-builder &&
+          doc-builder build datasets ../datasets/docs/source --library_name datasets --build_dir ../doc-build --clean --html &&
```
Comment (lhoestq) [inline .github/workflows/build_documentation.yml:79]: Do we need `library_name` here btw ? this is not present in `.github/workflows/build_dev_documentation.yml`

**Comment by stevhliu** [comment]
One more tiny error that doesn't seem specific to Datasets (Transformers example [here](https://huggingface.co/docs/transformers/multilingual#xlm-language-embeddings)), but apostrophes and symbols aren't properly displayed in the right navbar:

Screen Shot 2022-03-02 at 8 39 10 AM

**Comment by mishig25** [comment]
In the latest commit https://github.com/huggingface/datasets/pull/3690/commits/20bddf28b22798c309e6eb1198a716f055889e1b, I tried to reflect changes from https://github.com/huggingface/transformers/pull/15903 , however, the gh workflow is not being triggered. @lhoestq do you know why it might be the case?

**Comment by lhoestq** [comment]
I removed this line to trigger the job
```
  pull_request:
```

but got this error
```
[Error: .github#L1](https://github.com/huggingface/datasets/commit/033fe623c556b9dbc964708b672ff9bb4896c906#annotation_2897984435)
a step cannot have both the `uses` and `run` keys
```

**Comment by lhoestq** [comment]
It seems to be working again, let me re-add the line I removed

**Comment by mishig25** [comment]
@lhoestq if the CI passes, Im gonna merge this PR
please let me know if that sounds good
